README.md

```diff
@@ -53,7 +53,13 @@ import { computed } from '@ember/object';
 import { Ability } from 'ember-can';
 
 export default Ability.extend({
-  canWrite: computed('user.isAdmin', function() {
+  session: service(),
+
+  user: computed('session.currentUser', function() {
+    return this.get('session.currentUser');
+  }),
```

Comment:
If we want to show best practices these should be:

```js
user: reads('session.currentUser'),
canCreate: reads('user.isAdmin')
```

Reply:
We should even make it read-only:
```js
user: reads('session.currentUser').readOnly(),
canCreate: reads('user.isAdmin').readOnly()
```

But my purpose here was to explain the use case for everyone even for the beginners which may be not so familiar how the macros work exactly. Of course, the advanced developers will know how to improve it ;)

What do u think? Is it a good point?

Reply:
Why readOnly? What nefarious reason do we need to prevent a developer from setting a user?

If we attempt to test this and want to change the property of `canCreate` we cannot instead we would have to go into the source see that it is a read only property from user and attempt to set user oh but we can't because that too is read only so now we have to mock out session and set its currentUser with an isAdmin property all so we can manipulate canCreate. I fail to see the benifit for telling the world **never ever ever try to set this property to something you really think it should be**

The only use cases I've found for read only is in cases where the integrity of the class is dependent on the calculations done in a getter. For example in ember-data setting to the models' `attributes` would be problematic and so it makes sense it is a read only value. Where as in this case these are aliases not a derived values. They are not calculated, they are used.

Reply:
Keep in mind that this is just an example. But here are my thoughts.

Abilities shouldn't be overriden. That can break a lot. We can't have in one place the ability `canCreate` -> `true` and in the other `false` for the same model even if we want to. The `readOnly` will help to keep integrity between those states.

Second thing. We don't write code for easier testing. And even in tests, we shouldn't override the ability, just set the `isAdmin` or (more PRO) mock it using eg. Sinon.js.

I don't want to teach people here why they should (or not) use it, that's why I didn't add it. But as you know `reads` has some disadvantages and can mislead the expected behavior. More about it, eg. here: https://m.alphasights.com/the-risks-you-take-when-you-use-ember-computed-oneway-f2c008a14404

BTW. Im always making the property `readOnly` if I don't expect it to be ever overridden. Like using `const` and `let`. And I don't expect to override any ability on the fronted, as I'm not in charge to control what I can or not. That's a backend part. If you are creating abilities which are not dependent on the API and used only eg for display sth on UI, then you may want to set it on your own. In that case, there is no other choice than NOT to use `readOnly`.

Reply:
>Abilities shouldn't be overriden. That can break a lot. We can't have in one place the ability canCreate -> true and in the other false for the same model even if we want to. The readOnly will help to keep integrity between those states.

Ok, in this instance your argument has convinced me. Thank you.

>Im always making the property readOnly if I don't expect it to be ever overridden.

However, I do disagree with this. We have a dev on our team that is vehement about adding `readOnly` to everything and I've had many WTF moments and lost time trying to understand why things break. It has been my experience that it is better to be permissive then it is to introduce weird errors and unexpected crashes because someone somewhere felt they knew the future better then I. Personal opinion obviously.

I prefer to stick with the default for most cases and add a `readOnly` after I've convinced myself that changes to that property would be bad. As in your example above about abilities it is arguably true that changing an ability property manually would not be beneficial and `readOnly` makes sense there.

Thank you for the time to explain your perspective. It has helped me learn something new today.

Reply:
You are definitely right. We shouldn't add it everywhere we can. We have to be very careful with it. Of course, your experience will help you a lot with it, so with time, you will feel which are readOnly and which are not. The same as with consts. That requires some skill. But trust me there are some cases when this mark can save you a lot of trouble. You have to just use it wisely and exactly know the purpose. Not by guessing ;)

At the beginning of the Ember journey, I totally agree and do not recommend to use it until the moment you will realize that you really need it right now. Then you will be ready to control the force 😅

No problem. That's for the community is. Good to know your concerns as well :)

___
Anyway, you still think we should change the example to `reads`?

Reply:
I think the reads are good because most newbies don't realize they are there